### ISSUE TITLE
Allow IOutputLayers to feedforward if that IOutputLayer is not an output for the graph

### ISSUE THREAD

**Issue by edkek** [deeplearning4j/deeplearning4j#4146]
#### Issue Description

I know generative adversarial networks are not offically supported in DL4j, but I'm still attempting to make one in this framework anyways to see if it is at all possible.

I've hit a wall, where I can't specify a loss function for a layer unless that layer is an OutputLayer. This is fine normally, but if I have an OutputLayer be the input for another vertex/layer, then it will not feedforward, causing a `java.lang.IllegalStateException: Cannot do forward pass: all inputs not set`.

I'm not sure what causes this, I've made a gist [here](https://gist.github.com/edkek/be71116248f8c0682cb4fdfaf4aaa08f) where I can reproduce the bug consistently. (It just requires an `example.txt` file with whatever you want in it)

#### Version Information

Please indicate relevant versions, including, if relevant:

* Deeplearning4j version `0.8.0`
* platform information (OS, etc) `Windows 7 64-bit`
* CUDA version, if used `N/A`
* NVIDIA driver version, if in use `N/A`

#### Contributing

I would love to help fix this bug (I've already have the repo cloned and working), but I'm not sure where to start. I know it'd be in `ComputationGraph` (specifically somewhere [around here](https://github.com/deeplearning4j/deeplearning4j/blob/master/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/graph/ComputationGraph.java#L1474)). But some advice/guidance of where to start would help :)

**Comment by AlexDBlack** [comment]
In short, the reason for that design is that:
- Normal layer backprop is based on dL/dOut (i.e., activation gradients from layer above)
- Output layer backprop is based on labels

I haven't been keeping track of the progress in GAN architectures properly, but can you explain exactly what you are trying to do?
Specifically, what is the input to the loss function (other than the network activations, obviously) instead of a labels array?

**Comment by edkek** [comment]
Specifically, I was trying to implement these loss functions

<img width="347" alt="screen shot 2017-10-04 at 7 50 47 pm" src="https://user-images.githubusercontent.com/879484/31205131-80b82806-a93d-11e7-9cbf-ed2865b7f2da.png">

from [this paper](https://arxiv.org/abs/1611.09904)

I have two output layers
`gen_out` (G)
`dist_out` (D)

`gen_out` simply outputs it's generated data from noise, and it's loss function is dependent on whether the discriminator got it correct or not (see below). So it doesn't really need to be an output of the entire graph.

`dist_out` gets it's input from either `gen_out` or from a real input to the graph (it's chosen randomly during training), and it tries to output if the data is from a real input or from `gen_out`. 

I hope this helps

**Comment by edkek** [comment]
I figured out what I'm trying to do can be done with the `LossLayer`

Closing
